File: src/Scripts/C-Sharp/delete-current-file.cs

// C# Script that deletes the current file

// In input.conf add:

// KP0 script-message delete-current-file delete   #menu: Script > Delete current file > Delete
//   0 script-message delete-current-file delete   #menu: Script > Delete current file > Delete

// KP1 script-message delete-current-file confirm  #menu: Script > Delete current file > Confirm
//   1 script-message delete-current-file confirm  #menu: Script > Delete current file > Confirm

using System;
using System.IO;
using System.Threading;

using Microsoft.VisualBasic.FileIO;

using mpvnet;

class Script
{
    string FileToDelete;
    DateTime DeleteTime;
    CorePlayer Core;

    public Script()
    {
        Core = Global.Core;
        Core.ClientMessage += ClientMessage;
    }

    void ClientMessage(string[] args)
    {
        if (args == null || args.Length != 2 || args[0] != "delete-current-file")
            return;

        if (args[1] == "delete")
        {
            FileToDelete = Core.GetPropertyString("path");
            DeleteTime = DateTime.Now;
            Core.CommandV("show-text", "Press 1 to delete file", "10000");
        }
        else if (args[1] == "confirm")
        {
            TimeSpan ts = DateTime.Now - DeleteTime;
            string path = Core.GetPropertyString("path");

            if (FileToDelete == path && ts.TotalSeconds < 10 && File.Exists(FileToDelete))
            {
                Core.CommandV("show-text", "");

                int count = Core.GetPropertyInt("playlist-count");
                int pos = Core.GetPropertyInt("playlist-pos");
                int newPos = pos == count - 1 ? pos - 1 : pos + 1;

                if (newPos > -1)
                    Core.SetPropertyNumber("playlist-pos", newPos);

                Core.Command("playlist-remove " + pos);

                App.RunTask(() => {
                    Thread.Sleep(2000);
                    FileSystem.DeleteFile(FileToDelete, UIOption.OnlyErrorDialogs, RecycleOption.SendToRecycleBin);                
                });
            }
        }
    }
}

File: src/Scripts/Lua/delete-current-file.lua

-- Only supported on Windows.

-- This script deletes the file that is currently playing
-- via keyboard shortcut, the file is moved to the recycle bin.

-- Usage:
-- Configure input.conf as described below.
-- Press 0 to initiate the delete operation.
-- Press 1 to confirm and delete.

-- input.conf:

-- KP0 script-binding delete_current_file/delete
--   0 script-binding delete_current_file/delete

-- KP1 script-binding delete_current_file/confirm
--   1 script-binding delete_current_file/confirm

function delete()
    file_to_delete = mp.get_property("path")
    delete_time = os.time()
    mp.commandv("show-text", "Press 1 to delete file", "10000")
end

function confirm()
    local path = mp.get_property("path")

    if file_to_delete == path and (os.time() - delete_time) < 10 then
        mp.commandv("show-text", "")

        local count = mp.get_property_number("playlist-count")
        local pos   = mp.get_property_number("playlist-pos")
        local new_pos = 0

        if pos == count - 1 then
            new_pos = pos - 1
        else
            new_pos = pos + 1
        end

        if new_pos > -1 then
            mp.command("set pause no")
            mp.set_property_number("playlist-pos", new_pos)
        end

        mp.command("playlist-remove " .. pos)

        local ps_code = [[& {
            Start-Sleep -Seconds 2
            Add-Type -AssemblyName Microsoft.VisualBasic
            [Microsoft.VisualBasic.FileIO.FileSystem]::DeleteFile('file_to_delete', 'OnlyErrorDialogs', 'SendToRecycleBin')
        }]]

        local escaped_file_to_delete = string.gsub(file_to_delete, "'", "''")
        escaped_file_to_delete = string.gsub(escaped_file_to_delete, "’", "’’")
        escaped_file_to_delete = string.gsub(escaped_file_to_delete, "%%", "%%%%")
        ps_code = string.gsub(ps_code, "file_to_delete", escaped_file_to_delete)

        mp.command_native({
            name = "subprocess",
            playback_only = false,
            detach = true,
            args = { 'powershell', '-NoProfile', '-ExecutionPolicy', 'Bypass', '-Command', ps_code },
        })
    end
end

mp.add_key_binding(nil, "delete",  delete)
mp.add_key_binding(nil, "confirm", confirm)
